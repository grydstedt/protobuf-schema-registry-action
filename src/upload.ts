import fs from 'fs'
import protobuf, {ReflectionObject} from 'protobufjs'
import got from 'got'
import glob from 'glob'

interface ConfluentSchemaResponse {
  error_code: string
  message: string
}

export default async function upload(
  globPattern: string,
  url: string,
  username: string,
  password: string
): Promise<void> {
  const files = glob.sync(globPattern)

  for (const file of files) {
    const protoPlain = fs.readFileSync(file, 'utf8')
    const protoDescriptor = await protobuf.load(file)

    const messageWithTopic: ReflectionObject | undefined = Object.values(
      protoDescriptor.nested || {}
    ).find(msg => !!msg.options?.topic)

    if (!messageWithTopic)
      throw new Error('Protobuf does not a message with topic specified')

    const topic = messageWithTopic.options?.topic

    const {body}: {body: ConfluentSchemaResponse} = await got.post({
      url: `${url}/subjects/${topic}/versions`,
      headers: {
        'Content-Type': 'application/vnd.schemaregistry.v1+json',
        Authorization: `Basic ${Buffer.from(`${username}:${password}`).toString(
          'base64'
        )}`
      },
      throwHttpErrors: false,
      https: {rejectUnauthorized: false},

      body: JSON.stringify({
        schemaType: 'PROTOBUF',
        schema: protoPlain
      }),
      responseType: 'json'
    })
    if (body.error_code) throw new Error(body.message)
  }
}
