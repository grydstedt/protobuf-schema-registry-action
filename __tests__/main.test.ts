import upload from '../src/upload'
import got from 'got'

jest.mock('got')

const mockedGot = got as jest.Mocked<typeof got>

test('upload works', async () => {
  mockedGot.post.mockResolvedValue({body: {}})

  await upload(
    './__tests__/**/**.proto',
    'https://confluent-cloud',
    'U3ORYHY3WL3NRIOC',
    'ZMs277NGoy/P6LB6DCW1q6dg4CMl/292KinNsG1bMsuN9IIqeuw/62FEdGRcwfEA'
  )

  expect(mockedGot.post).toHaveBeenCalled()
  expect(mockedGot.post).toMatchSnapshot()
})
