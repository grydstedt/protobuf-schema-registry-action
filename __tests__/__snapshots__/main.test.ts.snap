// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload works 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "body": "{\\"schemaType\\":\\"PROTOBUF\\",\\"schema\\":\\"syntax = \\\\\\"proto3\\\\\\";\\\\n\\\\nmessage ExampleMessage {\\\\n\\\\n  option topic = \\\\\\"BusuuExampleEvent\\\\\\";\\\\n \\\\n  string message = 1;\\\\n  string extra = 2;\\\\n}\\"}",
        "headers": Object {
          "Authorization": "Basic VTNPUllIWTNXTDNOUklPQzpaTXMyNzdOR295L1A2TEI2RENXMXE2ZGc0Q01sLzI5Mktpbk5zRzFiTXN1TjlJSXFldXcvNjJGRWRHUmN3ZkVB",
          "Content-Type": "application/vnd.schemaregistry.v1+json",
        },
        "https": Object {
          "rejectUnauthorized": false,
        },
        "responseType": "json",
        "throwHttpErrors": false,
        "url": "https://confluent-cloud/subjects/BusuuExampleEvent/versions",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
